### PR TITLE
WooCommerce: Add currentlyEditing state for products and variations.

### PR DESCRIPTION
**Pull request by justinshreve**
Adds currentlyEditing state to both products and variations to track the current object being edited.

To test:

1. Run `make test` and ensure all tests succeed.

### CODE REVIEW

**Comment by allendav** [comment]
For this PR, I have kinda the same question as here: https://github.com/Automattic/wp-calypso/pull/13104/files#r111511067

**Comment by justinshreve** [comment]
@allendav In case you missed the reply, I've responded over at https://github.com/Automattic/wp-calypso/pull/13104#discussion_r111604347.

File: client/extensions/woocommerce/state/ui/products/edits-reducer.js [hunk 1]
```diff
@@ -29,9 +29,14 @@ function editProductAction( edits, action ) {
 
 	const prevEdits = edits || {};
 	const bucket = product && isNumber( product.id ) && 'updates' || 'creates';
-	const _array = editProduct( prevEdits[ bucket ], product, data );
+	const _product = product || { id: { index: ( prevEdits[ bucket ] || [] ).length } };
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/products/edits-reducer.js]: The internals of the reducers are starting to be a bit opaque - please consider well-named helper functions before expanding further.

File: client/extensions/woocommerce/state/ui/products/edits-reducer.js [hunk 2]
```diff
@@ -41,21 +46,25 @@ function editProductAttributeAction( edits, action ) {
 	const prevEdits = edits || {};
 	const bucket = product && isNumber( product.id ) && 'updates' || 'creates';
 	const _attributes = editProductAttribute( attributes, attribute, data );
-	const _array = editProduct( prevEdits[ bucket ], product, { attributes: _attributes } );
+	const _product = product || { id: { index: ( prevEdits[ bucket ] || [] ).length } };
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/products/edits-reducer.js]: Hmmm. In line with my comment above, could this be DRY-ed?

File: client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js [hunk 3]
```diff
@@ -27,24 +27,36 @@ function editProductVariationAction( edits, action ) {
...
-
 	let found = false;
 
 	// Look for an existing product edits first.
 	const _edits = prevEdits.map( ( productEdits ) => {
 		if ( productId === productEdits.productId ) {
 			found = true;
-			const _array = editProductVariation( productEdits[ bucket ], variation, data );
-			return { ...productEdits, [ bucket ]: _array };
+			const variationId = variation && variation.id || { index: ( productEdits[ bucket ] || [] ).length };
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js]: That pattern again :)

**Comment by justinshreve** [comment]
@allendav I've added a helpers file under `state/ui` so we can use the same pattern for other reducers. I've moved that code for figuring out the next bucket index, as well as the code for getting the updates or creates bucket, as that is repeated a few times as well. I'm sure other abstractions will make themselves obvious over time.  Does this look OK?

**Comment by allendav** [comment]
@justinshreve omg that's perfect :shipit:

**Pull request merged**